File: Problems/Algorithms/546. Number of Provinces/number_of_provinces.py
from collections import defaultdict


class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        n = len(isConnected)
        neighbors = defaultdict(list)
        for i in range(n):
            for j in range(i+1, n):
                if isConnected[i][j] == 1:
                    neighbors[i].append(j)
                    neighbors[j].append(i)

        visited = [False for _ in range(n)]

        def dfs(start: int, neighbors: Dict[int, List[int]], visited: List[int]):
            visited[start] = True

            for neighbor in neighbors[start]:
                if not visited[neighbor]:
                    dfs(neighbor, neighbors, visited)

        ans = 0
        for i in range(n):
            if not visited[i]:
                ans += 1
                dfs(i, neighbors, visited)

        return ans

File: Problems/Algorithms/546. Number of Provinces/number_of_provinces.go
func findCircleNum(isConnected [][]int) int {
    n := len(isConnected)
    neighbors := make(map[int][]int)
    for i := 0; i < n; i++ {
        for j := i+1; j < n; j++ {
            if isConnected[i][j] == 1 {
                neighbors[i] = append(neighbors[i], j)
                neighbors[j] = append(neighbors[j], i)
            }
        }
    }

    visited := make([]bool, n)

    ans := 0
    for i := 0; i < n; i++ {
        if !visited[i] {
            ans++
            dfs(i, neighbors, visited)
        }
    }

    return ans
}

func dfs(start int, neighbors map[int][]int, visited []bool) {
    visited[start] = true

    for _, neighbor := range neighbors[start] {
        if !visited[neighbor] {
            dfs(neighbor, neighbors, visited)
        }
    }
}
